### ISSUE TITLE
server: you can start two processes with the same --port as long as one specifies --listen-addr

### ISSUE THREAD

**Issue by andreimatei** [cockroachdb/cockroach#40708]
This has bit me several times already: `roachprod local` starts the first node as 
```
cockroach start --insecure [...] --listen-addr=127.0.0.1 --port=26257 --http-port=26258
```

This does not prevent you from starting another node as 
```
./cockroach start --insecure --logtostderr --port=26257
```
It does, however, prevent you from starting a node as
```
./cockroach start --insecure --logtostderr --port=26257 --listen-addr=127.0.0.1
```

This is most unfortunate, because on several occasions I had a `roachprod local` cluster running in the background, but then naively started another node for unrelated work, and tried to connect to it, only to connect to the wrong cluster.

I'm assuming node startup doesn't fail as long as there's any interfaces on which the port is available. But I think it should fail if the port is taken on *any* interface, and force you to specify a `listen-addr` if you really want to do something funky.

The code around this is very simple; we just call `net.Listen()`, which doesn't fail when I'd like it to fail.
https://github.com/cockroachdb/cockroach/blob/57ca9600603e788ec3f96f384fafe2821eb6a92f/pkg/server/server.go#L2230

@bdarnell , do you happen to have any suggestion here? I'd be happy to implement it.

**Comment by knz** [comment]
yes this is working as intended - having two processes, one listening on 0.0.0.0:26257 and the other on 127.0.0.1:26257, is valid.

**Comment by andreimatei** [comment]
> yes this is working as intended - having two processes, one listening on 0.0.0.0:26257 and the other on 127.0.0.1:26257, is valid.

It's a bit dismissive when two people who are not clueless are talking about an issue and the 3rd closes it by simply stating that it's working "as intended", because it begs the question of intended by whom? Also, you say it's intended and valid, but that's not how the system works on Linux - I've just tried and the second instance refuses to use the port. As Ben said, this seems to be a Mac only behavior.

**Comment by knz** [comment]
My point remains. It's a Macos thing **that is intended by apple**.

You would get the same behavior with `nc` or even a plain postgres server.

What do you think crdb should do about it?

**Comment by andreimatei** [comment]
> My point remains. It's a Macos thing that is intended by apple.

Well some low level behavior of some system call being "intended by Apple" does not generally translate into high-level CRDB behavior...

> You would get the same behavior with nc or even a plain postgres server.

I've just tried, and I don't believe this to be true for Postgres. I've configured one pg server to bind to all interfaces, and another one to bind to localhost. If they're both trying to use the same port, the second one doesn't start. It says that another Postmaster is already using `socket file /tmp/.s.PGSQL.5432`, and then in another part of the log it refers to that file as a "lock file".

> What do you think crdb should do about it?

One way or another, I would like for CRDB on OSX tro behave like it behaves on Linux (or at least how it appears to me to behave on Linux) - if I ask it to bind to all interfaces on port x, but it can't bind to some of them, I want it to fail.
Ben had a suggestion above about splitting the binding calls between the ipv6 and ipv4 interfaces. I think we should try that. Although it's not obvious to me that this is a ipv4 vs ipv6 thing, as opposed to more generally a single interface vs multiple interfaces thing.
If that doesn't work, we could try some lock files like pg apparently uses. PG does not appear to let you run two servers using the same port even if they're configured to bind on different interfaces. As far as I'm concerned, I could go either way on that.

**Comment by knz** [comment]
The behavior with "lock files" works with crdb already, if you let it configure a Unix socket. I remember ben was opposed to enabling the Unix socket by default, but maybe that would be a good use case for it.

**Comment by bdarnell** [comment]
I agree with Andrei that I don't think we can call this "working as intended". Regardless of apple's intentions here, our intention is that if you ask us to bind on all interfaces, you should get an error unless we were able to bind the requested port on all interfaces. But it's a really weird edge case - how likely are you to hit this situation without roachprod setting `--listen-addr` for you? I don't think it's worth changing the way we bind our ports to get a better error message in this case. There's a non-trivial risk of breaking something by doing that (I called out the issues around port 0, but there may be other things I'm not thinking of at the moment).

**Comment by knz** [comment]
Ben how do you feel about setting up the unix socket by default?

**Comment by bdarnell** [comment]
I don't understand how a unix socket helps here. Note that the invocations here must be running in different working directories because we're not running into store-level lock files, so where would we put the unix socket to do something useful here?

**Comment by knz** [comment]
The unix socket is typically created in `/tmp` or `/var/run` regardless of the directory you run your server in.
Also the unix socket is called `.s.PGSQL.NNNN` where NNNN is the port number.

This is exactly the functionality that makes PostgreSQL (and presumably mysql and other db servers) work the way that andrei wants on macos.

**Comment by bdarnell** [comment]
Unix sockets aren't great as lockfiles because they are often left behind if a process exits uncleanly. It's very common to blindly delete them (which is allowed, even if they are in use). If we turned on this socket automatically it would cause other headaches and we'd need to figure out a safe way to clean up. We shouldn't turn on unix sockets just because we want a lockfile.

**Comment by knz** [comment]
Ok I'm going to leave this issue open, but not touching this until we have a concrete, specific solution in mind with a clear analysis that it's not going to break existing behavior (e.g. the handling of auto-allocated port numbers).

**Comment by krishnakumarp** [comment]
I was stumped by this issue for a full day. I had installed Cockroach DB and had an instance of Cockroach running already and forgot about this.

`/opt/homebrew/opt/cockroach/bin/cockroach start-single-node --store=/opt/homebrew/var/cockroach --http-port=26256 --insecure --host=localhost`

Wrongly started a new instance 

`cockroach start-single-node --insecure`

After that I create a new database with some tables from a script file 

`cockroach sql --insecure --file <dbscript.sql>`


Now I have a spring/jpa app connecting to cockroach db using JDBC connection string 

`jdbc:postgresql://localhost:26257/mydb?sslmode=disable`

The application was reporting an error that it was not able to see the tables. 

The tables were visible from the command line interface though.

It took a long time to figure out that the root cause was that two instances were running.
